File: R/empcop.test.R
#################################################################
##   Copula R package by Jun Yan and Ivan Kojadinovic Copyright (C) 2007
##
##   Copyright (C) 2007 Ivan Kojadinovic <ivan@stat.auckland.ac.nz>
##
##   This program is free software; you can redistribute it and/or modify
##   it under the terms of the GNU General Public License as published by
##   the Free Software Foundation; either version 2 of the License, or
##   (at your option) any later version.
##
##   This program is distributed in the hope that it will be useful,
##   but WITHOUT ANY WARRANTY; without even the implied warranty of
##   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##   GNU General Public License for more details.
##
##   You should have received a copy of the GNU General Public License along
##   with this program; if not, write to the Free Software Foundation, Inc.,
##   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
##
##############################################################################

## Computes the sum of binomial coefficients
binom.sum <- function(n,k)
{
    bs <- 1
    for (i in 1:k)
        bs <- bs + choose(n,i)
    bs
}

##############################################################################

# simulate the distribution of TA for A of cardinality 2 to m under independence

empcop.simulate <- function(n,p,m,N=2000)
{
    if (!is.numeric(n) || as.integer(n) < 2)
        stop("n should be an integer greater than 2")
    if (!is.numeric(p) || as.integer(p) < 2)
        stop("p should be an integer greater than 2")
    if (!is.numeric(m) || as.integer(m) < 2 || as.integer(m) > as.integer(p))
        stop("m should be an integer greater than 2 and smaller than p")
    if (!is.numeric(N) || as.integer(N) < 100)
        stop("N should be an integer greater than 100")

    sb <- binom.sum(p,m)

    res<- .C("simulate_empirical_copula",
             n = as.integer(n),
             N = as.integer(N),
             p = as.integer(p),
             m = as.integer(m),
             TA0 = double(N * (sb - p - 1)),
             G0 = double(N),
             subsets = integer(sb),
             subsets.char = character(sb),
             PACKAGE="copula")

    d <- list(sample.size = n,
              data.dimension = p,
              max.card.subsets = m,
              number.repetitions = N,
              subsets = res$subsets.char[(p+2):sb],
              subsets.binary = res$subsets[(p+2):sb],
              dist.statistics.independence = matrix(res$TA0,N,sb-p-1),
              dist.global.statistic.independence = res$G0)
    
    class(d) <- "empcop.distribution"
           
    return(d)
}

#################################################################
##
##  Multivariate independence test based on the empirical 
##  copula process as proposed by Christian Genest and Bruno 
##  Rémillard (2004), Test 13:2, pages 335-369.
##
##  Ivan Kojadinovic, May 2007
##
##############################################################################


empcop.test <- function(x, d, m=ncol(x), alpha=0.05)
{
    if (!is.numeric(x <- as.matrix(x)))
        stop("data should be numerical")
    if (sum(is.na(x)) > 0)
        stop("data cannot contain missing values")
    if (nrow(x) < 2)
        stop("data should contain more than 2 rows")
    if (class(d) != "empcop.distribution")
        stop("d should be obtained by means of the function empcop.simulate")
    if (ncol(x) != d$data.dimension)
        stop("d was not obtained from simulations based on data whose dimension is equal to that of x")
    if (!is.numeric(m) || as.integer(m) < 2 || as.integer(m) > as.integer(d$max.card.subsets))
        stop(paste("m should be an integer greater than 2 and smaller than",d$max.card.subsets))
    if (!is.numeric(alpha) || alpha <= 0 || alpha >= 1)
        stop("the significance level alpha is not properly set")

    p <- ncol(x)
    N <- d$number.repetitions
      
    ## transform data to ranks
    for (j in 1:p)
        x[,j] <- rank(x[,j])

    sb <- binom.sum(p,m)
    
    ## perform test
    res<- .C("empirical_copula_test",
             as.integer(x),
             nrow(x),
             as.integer(p),
             as.integer(m),
             as.double(d$dist.statistics.independence),
             as.double(d$dist.global.statistic.independence),
             as.integer(N),
             as.integer(d$subsets.binary),
             TA = double(sb - p - 1),
             G = double(1),
             pval = double(sb - p - 1),
             fisher = double(1),
             tippett = double(1),
             globpval = double(1),
             PACKAGE="copula")
    
    ## compute critical values at the alpha level
    beta <- (1 - alpha)^(1 / (sb - p -1))
    critical <- numeric(0)
    from <- 1
    for (j in 2:m)
    {
        to <- from + choose(p,j) - 1 
        crit <- sort(as.double(d$dist.statistics.independence[,from:to]))[round(beta * N * (to - from + 1))]
        critical <- c(critical,rep(crit,choose(p,j)))
        from <- to + 1
    }
    
    test <- list(subsets=d$subsets[1:(sb - p - 1)],statistics=res$TA, critical.values=critical,
                 pvalues = res$pval, fisher.pvalue=res$fisher, tippett.pvalue=res$tippett, alpha=alpha,
                 beta=beta, global.statistic=res$G, global.statistic.pvalue=res$globpval)

    class(test) <- "empcop.test"
    return(test)
}

##############################################################################

dependogram <- function(test, pvalues=FALSE)
{
  if (class(test) != "empcop.test")
    stop("'test' should be obtained by means of the function empcop.test or empcop.rv.test")
  if (!is.logical(pvalues))
    stop("'pvalues' should be a boolean")

  par(las=3)
  
  l <- length(test$statistics)
  if (pvalues == TRUE)
    {
      plot(c(1,1:l),c(0,test$pvalues), type="h", ylab="p-value per subset", xlab="subsets",
           axes=FALSE, main="Dependogram")
      axis(1,at=1:l,label=test$subsets)
      axis(2)
      points(1:l,rep(1 - test$beta, l),pch=20)
    }
  else
    {
      plot(c(1,1:l),c(0,test$statistics), type="h", ylab="statistic per subset", xlab="subsets",
           axes=FALSE, main="Dependogram")
      axis(1,at=1:l,label=test$subsets)
      axis(2)
      points(1:l,test$critical.values,pch=20)
    }
}

##############################################################################
##
##  Independence test among random vectors based on the empirical 
##  copula process
##
##  Ivan Kojadinovic, December 2007
##
##############################################################################

empcop.rv.test <- function(x, d, m=length(d), N=1000, alpha=0.05)
{
    if (!is.numeric(x <- as.matrix(x)))
        stop("data should be numerical")
    if (sum(is.na(x)) > 0)
        stop("data cannot contain missing values")
    if ((n <- nrow(x)) < 2)
        stop("data should contain more than 2 rows")
    if (!is.numeric(d) || any(as.integer(d) < 1) || sum(d) != (nc <- ncol(x)))
      stop(paste("wrong vector of dimensions"))

    p <- length(d)
    
    if (!is.numeric(m) || as.integer(m) < 2 || as.integer(m) > p)
        stop(paste("m should be an integer greater than 2 and smaller than",p))
    if (!is.numeric(N) || as.integer(N) < 100)
        stop("N should be an integer greater than 100")
    if (!is.numeric(alpha) || alpha <= 0 || alpha >= 1)
        stop("the significance level alpha is not properly set")
      
    ## transform data to pseudo-observations
    for (j in 1:nc)
        x[,j] <- rank(x[,j])/n

    sb <- binom.sum(p,m)

    ## block indices
    b <- c(0,cumsum(d))
  
    ## bootstrap
    bootstrap <- .C("bootstrap",
                    n = as.integer(n),
                    N = as.integer(N),
                    p = as.integer(p),
                    b = as.integer(b),
                    U = as.double(x),
                    m = as.integer(m),
                    MA0 = double(N * (sb - p - 1)),
                    I0 = double(N),
                    subsets = integer(sb),
                    subsets.char = character(sb),
                    PACKAGE="copula")

    subsets <- bootstrap$subsets.char[(p+2):sb]
    subsets.binary <- bootstrap$subsets[(p+2):sb]
    dist.statistics.independence <- matrix(bootstrap$MA0,N,sb-p-1)

    ## perform test
    res<- .C("empirical_copula_test_rv",
             as.double(x),
             as.integer(n),
             as.integer(p),
             as.integer(b),
             as.integer(m),
             as.double(bootstrap$MA0),
             as.double(bootstrap$I0),
             as.integer(N),
             as.integer(subsets.binary),
             MA = double(sb - p - 1),
             I = double(1),
             pval = double(sb - p - 1),
             fisher = double(1),
             tippett = double(1),
             Ipval = double(1),
             PACKAGE="copula")
    
    ## compute critical values at the alpha level
    beta <- (1 - alpha)^(1 / (sb - p -1))
    critical <- numeric(0)
    sumcpj <- 0
    for (j in 2:m)
      {
        cpj <- choose(p,j)
        for (k in 1:cpj)
          {
            crit <- sort(dist.statistics.independence[,sumcpj + k])[round(beta * N)]
            critical <- c(critical,crit)
          }
        sumcpj <- sumcpj + cpj
      }
    
    test <- list(subsets=subsets[1:(sb - p - 1)],statistics=res$MA, critical.values=critical,
                 pvalues = res$pval, fisher.pvalue=res$fisher, tippett.pvalue=res$tippett, alpha=alpha,
                 beta=beta, global.statistic=res$I, global.statistic.pvalue=res$Ipval)

    class(test) <- "empcop.test"

    return(test)
}

##############################################################################

File: src/empcop.test.c
/*#################################################################
##   Copula R package by Jun Yan and Ivan Kojadinovic Copyright (C) 2007
##
##   Copyright (C) 2007 Ivan Kojadinovic <ivan@stat.auckland.ac.nz>
##
##   This program is free software; you can redistribute it and/or modify
##   it under the terms of the GNU General Public License as published by
##   the Free Software Foundation; either version 2 of the License, or
##   (at your option) any later version.
##
##   This program is distributed in the hope that it will be useful,
##   but WITHOUT ANY WARRANTY; without even the implied warranty of
##   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##   GNU General Public License for more details.
##
##   You should have received a copy of the GNU General Public License along
##   with this program; if not, write to the Free Software Foundation, Inc.,
##   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
##
#################################################################*/

/*****************************************************************************

  Multivariate independence test  based on the empirical 
  copula process as proposed by Christian Genest and Bruno 
  Rémillard (2004), Test 13:2, pages 335-369.

  Ivan Kojadinovic, May 2007, modified Dec 2007

*****************************************************************************/

#include <R.h>
#include <Rmath.h>
#include "set.utils.h"

/*****************************************************************************

  Computation of the Cramer-von Mises statistics used in the independence
  test of Genest and Remillard based on the empirical copula for a
  subset A of {1,...,p}
  
******************************************************************************/

double global_stat(int n, int p, int *R)
{
  int i,j,l;
  double stat, sum, prod, part1, part2;
  
  /* first term */
  sum = 0.0;
  for (i=0;i<n;i++)
    for (l=0;l<n;l++) 
      {
	prod = 1.0;
	for (j=0;j<p;j++)
	  prod *= 1.0 - fmax2(R[i + n*j], R[l + n*j])/n;
	sum += prod;
      }
  part1 = sum/n;
  
  /* second term */
  sum = 0.0;
  for (i=0;i<n;i++)
    {
      prod = 1.0;
      for (j=0;j<p;j++)
	prod *= (n * (n - 1.0) - R[i + n*j] * (R[i + n*j] - 1.0) )
	  /(2.0 * n * n);
      sum += prod;
    }
  part2 = sum;

  stat = part1 - 2.0 * part2 
    + n * R_pow_di((n - 1.0) * (2.0 * n - 1.0) / (6.0 * n * n) ,p);

  return stat;
}

double Dn(int n, int s, int t)
{
  return (2.0 * n + 1.0) * (n + 1.0) / (6.0 * n * n) 
    + (s * (s - 1.0) + t * (t - 1.0)) / (2.0 * n * n) 
    - fmax2(s,t) / n;
}

double empirical_copula_subset(int n, int p, int *R, int A)
{
  int i,j,k;
  double TA = 0.0, prod;
  
  for (i=0;i<n;i++)
    for (k=0;k<n;k++) {
      prod = 1.0;
      for (j=0;j<p;j++)
	if (1<<j & A)
	  prod *= Dn(n, R[i + n*j], R[k + n*j]);
      TA += prod;
    }
  return TA/n;
}

/*****************************************************************************

  Simulate the distribution of TA, up to subsets of cardinality p
  and of the global statistic
  n: sample size
  N: number of repetitions
  p: dimension of data
  m: max. card. of A
  TA0: values of TA under independence (N repetitions) 
  G0: values of the global stat. under independence (N repetitions)
  subset: subsets of {1,...,p} in binary notation (int) whose card. is 
  between 2 and m in "natural" order
  subset_char: similar, for printing
  
******************************************************************************/

void simulate_empirical_copula(int *n, int *N, int *p, int *m, double *TA0, 
			       double *G0, int *subset, char **subset_char)
{
  int i, j, k, t1, t2, sb = (int)sum_binom(*p,*m);
  int *R = (int *)R_alloc((*n) * (*p), sizeof(int));
  
  /* partial power set in "natural" order */ 
  k_power_set(p, m, subset);

  /* convert partial power set to char for printing */
  k_power_set_char(p,m, subset, subset_char);
  
  /* Simulate the distribution of TA under independence */

  GetRNGstate();
  
  /* N repetitions */
  for (k=0;k<*N;k++) { 

    Rprintf("Simulation iteration %d\n",k+1);
    
    /* generate data */
    for (j=0;j<*p;j++) 
      {	
	for (i=0;i<*n;i++)
	  R[i + (*n) * j] = i+1;
	
	/* permutation = random ranks in column j */
	for (i=*n-1;i>=0;i--)
	  {
	    t1 = R[j * (*n) + i]; 
	    t2 = (int)(i *  unif_rand());
	    R[j * (*n) + i] = R[j * (*n) + t2];
	    R[j * (*n) + t2] = t1;
	  }
      }
    
    /* for subsets i of cardinality greater than 1 and lower than m */
    for (i=*p+1;i<sb;i++) 
      TA0[k + (*N) * (i - *p - 1)] = empirical_copula_subset(*n, *p, R, subset[i]);

    /* global stat under independence*/
    G0[k] = global_stat(*n, *p, R);
  }
  PutRNGstate();
}

/*****************************************************************************

  Compute TA for subsets of cardinality 2 to p
  R: ranks
  n: sample size
  p: number of variables
  m: max. cardinality of subsets of {1,...,p}
  TA0: simulated values of TA under independence (size: N * (sb - p - 1))
  N: number of repetitions (nrows TA0, fisher0, tippett0)
  subset: subsets of {1,...,p} in binary notation (int) whose card. is 
  between 2 and m in "natural" order
  TA: test statistics (size: sum.bin. - p - 1) 
  pval: corresponding p-values (size = sum.bin. - p - 1)
  fisher: pvalue à la Fisher
  tippett: pvalue à la Tippett
  
******************************************************************************/

void empirical_copula_test(int *R, int *n, int *p, int *m, double *TA0, double *G0,
			   int *N, int *subset, double *TA, double *G, double *pval,
			   double *fisher, double *tippett, double *globpval)			  
{
  int i, j, k, count, sb = (int)sum_binom(*p,*m);
  double *fisher0 = (double *)R_alloc(*N, sizeof(double));
  double *tippett0 = (double *)R_alloc(*N, sizeof(double));
  double pvalue;

  /* compute W à la Fisher and à la Tippett from TA0*/ 
  for (k=0;k<*N;k++) {
    fisher0[k] = 0.0;
    tippett0[k] = 1.0;
    for (i=*p+1;i<sb;i++) {
      /* p-value */
      count = 1;
      for (j=0;j<*N;j++) 
	if (TA0[j + (*N) * (i - *p - 1)] > TA0[k + (*N) * (i - *p - 1)])
	  count ++;
      pvalue = (double)(count + 0.5)/(*N + 1.0);
      fisher0[k] -= 2*log(pvalue);
      tippett0[k] = fmin2(tippett0[k],pvalue);
    }
  }

  /* compute W from the current data */
  *fisher = 0.0;
  *tippett = 1.0;
  
  /* for subsets i of cardinality greater than 1 */
  for (i=*p+1;i<sb;i++) 
    {
      TA[i - *p - 1] = empirical_copula_subset(*n, *p, R, subset[i - *p - 1]);
      
      /* p-value */
      count = 1;
      for (k=0;k<*N;k++) 
	if (TA0[k + (*N) * (i - *p - 1)] > TA[i - *p - 1])
	  count ++;
      pval[i - *p - 1] = (double)(count + 0.5)/(*N + 1.0);
      
      *fisher -= 2*log(pval[i - *p - 1]);
      *tippett = fmin2(*tippett,pval[i - *p - 1]);
    }

  /* p-values of the Fisher and Tippett statistics */
  count = 1;
  for (k=0;k<*N;k++) 
    if (fisher0[k] > *fisher)
      count ++;
  *fisher = (double)(count)/(*N + 1.0);

  count = 1;
  for (k=0;k<*N;k++) 
    if (tippett0[k] < *tippett)
      count ++;
  *tippett = (double)(count)/(*N + 1.0);


  /* compute global stat from the current data and the corresponding p-value*/ 
  *G = global_stat(*n, *p, R); 
  count = 1;
  for (k=0;k<*N;k++) 
    if (G0[k] > *G)
      count ++;
  *globpval = (double)(count)/(*N + 1.0);
}


